sql/moz-fx-data-shared-prod/telemetry_derived/firefox_android_clients_v1/query.sql

```diff
@@ -0,0 +1,482 @@
+        NULL,
+        NULL,
+        metrics.install_source,
+        metrics.metrics_ping_submission_date
+      ) AS install_source__source_ping_date
+    ) AS metadata
+  FROM
+    existing
+  LEFT JOIN
+    first_session_ping first_session
+  USING
+    (client_id)
+  LEFT JOIN
+    metrics_ping_unique AS metrics
+  USING
+    (client_id)
+),
+first_seen_with_pings_new_data AS (
+  SELECT
+    first_seen.client_id,
+    first_seen.first_seen_date,
+    first_session.first_run_date AS first_run_date,
+    first_session.first_reported_country AS first_reported_country,
+    first_session.first_reported_isp AS first_reported_isp,
+    first_session.channel AS channel,
+    first_session.device_manufacturer AS device_manufacturer,
+    first_session.device_model AS device_model,
+    first_session.os_version AS os_version,
+    first_session.adjust_campaign AS adjust_campaign,
+    first_session.adjust_ad_group AS adjust_ad_group,
+    first_session.adjust_creative AS adjust_creative,
+    COALESCE(first_session.adjust_network, metrics.adjust_network) AS adjust_network,
+    metrics.install_source AS install_source,
+    STRUCT(
+      CASE
+      WHEN
+        first_session.client_id IS NULL
+      THEN
+        FALSE
+      ELSE
+        TRUE
+      END
+      AS reported_first_session_ping,
+      CASE
+      WHEN
+        metrics.client_id IS NULL
+      THEN
+        FALSE
+      ELSE
+        TRUE
+      END
+      AS reported_metrics_ping,
+      first_session.first_run_date AS min_first_session_ping_date,
+      CASE
+      WHEN
+        first_session.adjust_network IS NOT NULL
+      THEN
+        'first_session'
+      WHEN
+        first_session.adjust_network IS NULL
+        AND metrics.adjust_network IS NOT NULL
+      THEN
+        'metrics'
+      ELSE
+        NULL
+      END
+      AS adjust_network__source_ping,
+      'metrics' AS install_source__source_ping,
+      CASE
+      WHEN
+        first_session.adjust_network IS NOT NULL
+      THEN
+        first_session.first_run_date
+      WHEN
+        first_session.adjust_network IS NULL
+        AND metrics.adjust_network IS NOT NULL
+      THEN
+        metrics.metrics_ping_submission_date
+      ELSE
+        NULL
+      END
+      AS adjust_network__source_ping_date,
+      CASE
+      WHEN
+        metrics.install_source IS NOT NULL
+      THEN
+        metrics.metrics_ping_submission_date
+      ELSE
+        NULL
+      END
+      AS install_source__source_ping_date
+    ) AS metadata
+  FROM
+    first_seen
+  LEFT JOIN
+    first_session_ping first_session
+  USING
+    (client_id)
+  LEFT JOIN
+    metrics_ping_unique AS metrics
+  USING
+    (client_id)
+)
+SELECT
+  *
+FROM
+  existing_with_pings_new_data
+UNION ALL
+SELECT
+  first_seen_with_pings_new_data.*
+FROM
+  first_seen_with_pings_new_data
+LEFT JOIN
+  existing_with_pings_new_data
+USING
+  (client_id)
+WHERE
+  existing_with_pings_new_data.client_id IS NULL
```

Comment:
I'm not following what's going on here. Why are we unioning then joining, if the client_id is going to be null after the join (and that is the join key?)

Reply:
This condition avoids duplicated records for existing clients.
E.g. If `client1` is already in the table and there is a new valid attribution on @submission_date, the record is updated in `first_seen_with_pings_new_data` but it will also appear in `first_seen_with_pings_new_data`, so it would appear twice in the result.
This line excludes the existing records from the `first_seen_with_pings_new_data` set.

Reply:
I see. You're essentially recreating a `FULL OUTER JOIN` that could encompass both of the CTEs (`first_seen_with_pings_new_data` and `existing_pings_with_new_data`). What do you think about switching this to an outer join and handling the three possibilities there?
- Existing data, but no new data
- Existing and new data
- New data, but no existing data